src/engine/strat_engine/thinpool/thinpool.rs

```diff
@@ -1058,181 +952,126 @@ impl ThinPool {
-                );
-                let (data_segments, meta_and_spare_segments) = (
-                    data_segment.map(|seg| coalesce_segs(data_existing_segments, &[seg])),
-                    meta_segments.map(|(seg, seg_s)| {
+            let device = backstore
+                .device()
+                .expect("If request succeeded, backstore must have cap device.");
+
+            let requests = vec![meta_extend_size, meta_extend_size];
+            let meta_index = 0;
+            let spare_index = 1;
+            match backstore.request_alloc(&requests) {
+                Ok(Some(transaction)) => {
+                    let backstore_segs = transaction.get_backstore();
+                    backstore.commit_alloc(pool_uuid, transaction)?;
+
+                    let meta_and_spare_segment = backstore_segs.get(meta_index).and_then(|seg| {
+                        backstore_segs.get(spare_index).map(|seg_s| (*seg, *seg_s))
+                    });
+                    let meta_and_spare_segments = meta_and_spare_segment.map(|(seg, seg_s)| {
                         (
                             coalesce_segs(meta_existing_segments, &[seg]),
                             coalesce_segs(spare_meta_existing_segments, &[seg_s]),
                         )
-                    }),
-                );
-
-                // Meta extension must be done first because growing the data device
-                // first could cause inadequate metadata space if the metadata
-                // extension fails.
-                if let Some((mut ms, mut sms)) = meta_and_spare_segments {
-                    // Leaves meta device suspended
-                    thinpooldev.set_meta_table(get_dm(), segs_to_table(device, &ms))?;
-
-                    meta_existing_segments.clear();
-                    meta_existing_segments.append(&mut ms);
+                    });
 
-                    spare_meta_existing_segments.clear();
-                    spare_meta_existing_segments.append(&mut sms);
-                }
+                    if let Some((mut ms, mut sms)) = meta_and_spare_segments {
+                        // Leaves meta device suspended
+                        thinpooldev.set_meta_table(get_dm(), segs_to_table(device, &ms))?;
 
-                if let Some(mut ds) = data_segments {
-                    // Leaves data device suspended
-                    thinpooldev.set_data_table(get_dm(), segs_to_table(device, &ds))?;
+                        meta_existing_segments.clear();
+                        meta_existing_segments.append(&mut ms);
 
-                    data_existing_segments.clear();
-                    data_existing_segments.append(&mut ds);
-                }
+                        spare_meta_existing_segments.clear();
+                        spare_meta_existing_segments.append(&mut sms);
+                    }
 
-                thinpooldev.resume(get_dm())?;
+                    if let Some((seg, _)) = meta_and_spare_segment {
```

Comment:
This is actually an interesting one because it looks like the refactor causes the thin pool to be suspended twice instead of just once. Do you have any reservations about that? Do you, @drckeefe?

Reply:
I believe that suspending twice is fine.

Reply:
Okay great!